Add mediators for preparing a retro

joshuatobin opened the pull request:
This PR adds a few new mediators for preparing for a retro by introducing support for the google api execution service. Still working on tests but opening early since @lexelby is out next week.

CC: @reidmix

lexelby commented on app/admin/incident.rb:
```diff
@@ -16,26 +18,29 @@
-        title:               resource[:title],
-        trello_oauth_token:  current_user.trello_oauth_token,
-        trello_oauth_secret: current_user.trello_oauth_secret
+      incident = Incident.find_by(id: resource[:id])
```
`resource` here should actually be the Incident.  You could just pass it in instead.

joshuatobin replied:
Not sure what you mean here? I had to do the `find_by` to get it to work. Maybe there is a better way but I couldn't seem to just pass in `resource` if thats what you mean?

joshuatobin replied:
Nevermind. Im not sure why it wasn't working originally. Fixed.

lexelby commented on app/admin/incident.rb:
```diff
@@ -16,26 +18,29 @@
-      session[:return_to] = resource[:incident_id]
+      session[:return_to] = resource[:id]
       redirect_to "/auth/trello"
+    rescue Signet::AuthorizationError, Google::Apis::AuthorizationError
```
I'd rather this code didn't have to know the specific Trello and Google exceptions that are thrown by their respective gems.  The mediator is where the logic about deciding what to do with those gems' exceptions should be.

You could have the mediator declare two exception classes, `TrelloAuthRequired` and `GoogleAuthRequired`, and just catch those here.

lexelby commented on app/admin/incident.rb:
```diff
@@ -45,8 +50,8 @@
-  action_item :create_trello_card, only: :post do
-    link_to 'Create Trello Card', create_trello_card_admin_incident_path(resource)
+  action_item :prepare_retro, only: :post do
+    link_to 'Prepare Retro Card/Doc', prepare_retro_admin_incident_path(resource)
```
Is there a reason this one is a GET whereas the one below is a POST?  IMO they should both be POST.

joshuatobin replied:
Fixed.

reidmix replied:
FWIW, another reason they should favor PUT/POST is because they are actions that create a change in the system.  For example, browsers that have prefetch functionality could accidentally create the trello/gdoc if they were to traverse the link.

lexelby commented on app/controllers/trello_controller.rb:
```diff
@@ -8,9 +8,12 @@ def create
+    )
 
-    redirect_to create_trello_card_admin_incident_path(session.delete(:return_to))
+    redirect_to prepare_retro_admin_incident_path(session.delete(:return_to)
```
missing close paren

joshuatobin replied:
Missing a `)` here.

lexelby commented on app/support/mediators/incident/create_card.rb:
```diff
@@ -18,11 +18,14 @@ def call
+        "keepFromSource" => "all"
       )
+
+      @card
```
Why are you saving this as an instance variable now?

joshuatobin replied:
This needs to return `id` and `url`

lexelby replied:
Right, but why not just have `@trello.create(...` be the last statement so the card is returned directly?  Why save it to an instance variable and then return the variable?

lexelby commented on app/support/mediators/incident/edit_card.rb:
```diff
@@ -0,0 +1,34 @@
+      raise
+    end
+
+    def update_description
```
The old one does a search and replace using the existing card description, rather than replacing it entirely.

joshuatobin replied:
/shrug. This was easier and is cleaner IMO.

lexelby replied:
...but we're probably ripping out trello anyway.  Ignore me.

lexelby commented on app/support/mediators/incident/prepare_retro.rb:
```diff
@@ -0,0 +1,49 @@
+      @incident, @current_user = incident, current_user
+    end
+
+    def call
```
I like how clean this method became, just calling out to other mediators.  Nice work.

joshuatobin replied:
👍

reidmix replied:
agreed, it's super clean.

lexelby commented on config/config.rb:
```diff
@@ -40,4 +40,8 @@ module Config
   optional :trello_template, string
+
+  # Google Script
+  optional :google_script_id, string
```
Could you add a comment indicating what this script is for, and perhaps also check the script itself into the repo somewhere?

joshuatobin replied:
I don't think we want to check the script into a public repo.

lexelby replied:
Well, we already have the retrospective template itself posted publicly...

joshuatobin commented on app/support/mediators/incident/create_card.rb:
```diff
@@ -18,11 +18,14 @@ def call
+        "name"           => "Incident #{@id}: #{@title}",
+        "idCardSource"   => @template.id,
+        "idList"         => @template.list_id,
+        "keepFromSource" => "all"
```
Even though the docs and code say this is the default, I still had to add it to get the checklists to copy over.

joshuatobin commented on app/support/mediators/incident/create_retro_doc.rb:
```diff
@@ -0,0 +1,27 @@
+module Mediators::Incident
+  class CreateRetroDoc < Mediators::Base
+
+    def initialize(auth:, id:, title:, trello_url:, postmortem_date: false, dev_mode: false)
```
I'd like to just have this take `**args` but for another time.

joshuatobin commented on app/support/mediators/incident/edit_card.rb:
```diff
@@ -0,0 +1,34 @@
+
+    def update_description
+      "[Heroku status incident](https://status.heroku.com/admin/incidents/#{@id})\n\n" \
+      "[Gdoc notes](#{@gdoc})\n\n"
```
Hard coding this but its going to get ripped out eventually as the trello integration is somewhat useless.

lexelby replied:
True.

joshuatobin commented on app/support/mediators/incident/prepare_retro.rb:
```diff
@@ -0,0 +1,52 @@
+class GoogleAuthRequired < StandardError; end
+class TrelloAuthRequired < StandardError; end
+
```
@lexelby New exception classes

reidmix replied:
These are also defined in `app/admin/incident.rb` -- you will likely get "constant already defined" warnings. 
And it will be better to have them in one place.

joshuatobin commented:
@lexelby I believe I've addressed all your concerns. 

@reidmix can you take a look and let me know what you'd like to see in specs to get this merged down?

reidmix commented on app/support/mediators/incident/create_retro_doc.rb:
```diff
@@ -0,0 +1,30 @@
+      @auth, @id, @title, @trello_url, @postmortem_date, @dev_mode = auth, id, title, trello_url, postmortem_date, dev_mode
+
+      @service = Google::Apis::ScriptV1::ScriptService.new
+      @service.authorization = @auth
```
you can keep `auth` a local variable here since it's not used outside of the method
ie. `auth` instead of `@auth`

joshuatobin replied:
👍 fixed

lexelby commented:
Yup, my concerns look good now.  Just need to fix what reid pointed out :)

lexelby commented:
:+1:
